### ISSUE TITLE
Need to know API call with promise return

### ISSUE THREAD

**Issue by InterestedLearner** [iroy2000/react-redux-boilerplate#51]
Hi I have started my first project with this structure, I am trying to call a method from Component and want the promise return there. As per the structure I am able to call an API and getting the result in a variable. All I want to return the promise. here is my sample code:-

a. On componentDidMount I have called a method of Mudules to create an Action.
`// On view`
` componentDidMount() {`
     `this.props.getCampaignTopics(1);`
` }`

 `//On Modules`
 `export const getCampaignTopics = createAction(GET_EXAMPLES, (id : string) => ({id}));`

b. On saga when action has been fired, I am calling the API endpoint to get the data
`// On Saga`
`export function* getCampaignTopics(actions) {`
  ` let id=0;`
   `if(actions.payload.id)`
   `{`
       `id= actions.payload.id;`
  ` }`
   `const responsetype = yield call(() => fetch(__CONFIG__.api_url+'example/' + id,{method:'GET',data:{}}).then(response => response.json()))`

  `yield put(addcampaignActions.setCampaignTopics(responsetype.data));`
 ` }`

`function* watchGetHome() {`
   `yield takeLatest(addcampaignConstants.GET_EXAMPLES, getCampaignTopics);`
`}`

c. After Getting the data I am calling another method of Modules to set data
`export const setCampaignTopics = createAction(SET_EXAMPLES, (topics : any) => ({topics}));`

`export const reducers = {`
  ` [GET_EXAMPLES]: (state, { payload }) =>`
    `state.merge({`
     ` ...payload,`
    `}),`
     ` [SET_EXAMPLES]: (state, { payload }) =>`
       ` state.merge({`
          `...payload,`
       ` }),`
    `}`

d. In View I am using selector to separate the indexes
`// Selector`
`const exampleSelector = state => state.addcampaign;`

`const topicSelector = createSelector(`
  `exampleSelector,`
  `payload => payload.get('topics')`
`);`

`export const exampleSelector = state => ({`
 ` topics: topicSelector(state)`
`});``

e. In component I am getting the value on props
`// render on Component
const { addcampaign } = this.props;     
topics = addcampaign.topics.toJS().children;`

Please tell me the right process to call an API and get result. Can anyone please tell me how can I get the result on promise. Here is how I want:-
`this.props.getCampaignTopics(1).then((result)=>{
// Need result here
});`

**Comment by iroy2000** [comment]
Let me rephrase the question, let me know if I understand your question correctly.

So instead of going through "indirect" layer of "calling action, saga, call api, calling action to set return value, use selector in the view",  you just want in the view directly talk to the api and get the results ?

If that's what you want, why not just create an entity object and wrap the api call?

```
// entity/campaign.js

export function getCampaignTopics(id) {
  return fetch(__CONFIG__.api_url+'example/' + id,{method:'GET',data:{}})
}

.... and some where in your View

import { getCampaignTopics } from 'entity/campaign'

getCampaignTopics(1).then(result => ...)

```

**Comment by InterestedLearner** [comment]
Thanks Roy for your answer. I am trying to learn saga that's all. Yes you understood my question, so as per your answer I can do that using an entity object. But can't I do that using saga middleware? If I can't then what is the right process to do that using saga. As for example, I have an edit profile page where I will get the user details on componentDidMount but I am getting the return values(Using indirect ways) in render method, which I think not a right way to do that and after submitting the data I am getting retrun again in render. So I have check every time if variable has some value with success identifier and failure identifier. Could you please describe me something so it is clear for me why I should use indirect method and what is the right way to do the edit profile process using Redux-Saga?

**Comment by iroy2000** [comment]
Redux Saga is for managing async `side effect` in `synchronize` manner by using generator.  And also, think Redux Saga as a process manager for your redux actions. If your use case is simple, you might not see too much benefit from it, but if you have complicated UI ( multiple operation ), you will see Saga makes your life easier. 

For your profile case, your code seems fine, but think about the following case

User make changes to their profile --> update your store ( happy path update ) --> calling the api --> update your store again ( to make sure it sync ) --> ( what if the api failed, handle error )  -->   update sidebar because user change profile.

ReduxSaga really fly if you need to handle complicated sequence of event in the same function.
